test/workbox-runtime-caching/node/test-CacheFirst.mjs

```diff
@@ -0,0 +1,169 @@
+  });
+
+  it(`should be able to cache a non-existant request to custom cache`, async function() {
+    const cacheName = 'test-cache-name';
+    const request = new Request('http://example.io/test/');
+    // Doesn't follow spec: https://github.com/pinterest/service-workers/issues/52
+    const event = new FetchEvent(request);
+
+    sandbox.stub(global, 'fetch').callsFake((req) => {
+      expect(req).to.equal(request);
+      return Promise.resolve(new Response('Hello Test.'));
+    });
+    let cachePromise;
+    sandbox.stub(event, 'waitUntil').callsFake((promise) => {
+      cachePromise = promise;
+    });
+
+    const cacheFirst = new CacheFirst({
+      cacheName,
+    });
+    const firstHandleResponse = await cacheFirst.handle(event);
+
+    // Wait until cache.put is finished.
+    await cachePromise;
+    const cache = await caches.open(cacheName);
+    const firstCachedResponse = await cache.match(request);
+
+    await compareResponses(firstHandleResponse, firstCachedResponse, true);
+  });
+
+  it(`should not cache an opaque response by default`, async function() {
+    const request = new Request('http://example.io/test/');
+    // Doesn't follow spec: https://github.com/pinterest/service-workers/issues/52
+    const event = new FetchEvent(request);
+
+    sandbox.stub(global, 'fetch').callsFake((req) => {
+      expect(req).to.equal(request);
+      return Promise.resolve(new Response('Hello Test.', {
+        status: 0,
+      }));
+    });
+    let cachePromise;
+    sandbox.stub(event, 'waitUntil').callsFake((promise) => {
+      cachePromise = promise;
+    });
+
+    const cacheFirst = new CacheFirst();
+    const firstHandleResponse = await cacheFirst.handle(event);
+
+    // Wait until cache.put is finished.
+    await cachePromise;
+    const cache = await caches.open(_private.cacheNames.getRuntimeName());
+    const firstCachedResponse = await cache.match(request);
+
+    expect(firstCachedResponse).to.equal(null);
+    expect(firstHandleResponse).to.exist;
+  });
+
+  it(`should cache an opaque response when a cacheWillUpdate plugin returns true`, async function() {
+    const request = new Request('http://example.io/test/');
+    // Doesn't follow spec: https://github.com/pinterest/service-workers/issues/52
+    const event = new FetchEvent(request);
+
+    sandbox.stub(global, 'fetch').callsFake((req) => {
+      expect(req).to.equal(request);
+      return Promise.resolve(new Response('Hello Test.', {
+        status: 0,
+      }));
+    });
+    let cachePromise;
+    sandbox.stub(event, 'waitUntil').callsFake((promise) => {
+      cachePromise = promise;
+    });
+
+    const cacheFirst = new CacheFirst({
+      plugins: [
+        {
+          cacheWillUpdate: ({request, response}) => {
+            return response;
+          },
+        },
+      ],
+    });
+    const firstHandleResponse = await cacheFirst.handle(event);
+
+    // Wait until cache.put is finished.
+    await cachePromise;
```

Comment:
Nice use of the `event.waitUntil()` stub to detect when the caching is complete!